src/spatial/spatial_tree.h

```diff
@@ -0,0 +1,225 @@
+pair<pair<point, T>, pair<point, T> > MakeNeighborPair(pair<point, T> p1,
+                                                       pair<point, T> p2) {
+  return make_pair(make_pair(p1.first, p1.second),
+                   make_pair(p2.first, p2.second));
+}
+
+/**
+ * Neighbour search function
+ * Search pairs (a, b) where a from A and b from B
+ * @tparam T
+ * @param A - node of the tree
+ * @param B - node of the tree
+ * @param distance - finding neighbors within that distance
+ * @param result - container, where we save retrieved data (all neighbor pairs
+ * with its points)
+ */
+template <typename T>
+void spatial_tree_node<T>::GetNeighbors(
+        spatial_tree_node<T> *A, spatial_tree_node<T> *B, double distance,
+        vector<pair<pair<point, T>, pair<point, T> > > *result) {
+  // Two cases:
+  //   A and B are leaf nodes - check every pair
+  //   otherwise do recoursive call on children
+  if (A->IsLeaf() && B->IsLeaf()) {
+    auto a_objs = A->GetObjects();
+    auto b_objs = B->GetObjects();
+
+    bool is_same = A == B;
+
+    for (int i = 0; i < a_objs->size(); i++)
+      for (int j = is_same ? (i + 1) : 0; j < b_objs->size(); j++)
+        if ((a_objs->at(i).first).SquaredEuclidianDistance(b_objs->at(j).first) <= distance)
+          result->push_back(MakeNeighborPair(a_objs->at(i), b_objs->at(j)));
+  } else {
+    spatial_tree_node<T> **a_nodes, **b_nodes;
+    int a_size = 0, b_size = 0;
+    // If A is a leaf node than call with A and child of B
+    if (A->IsLeaf()) {
+      b_nodes = B->GetChildrenNodes();
+      for (int i = 0; i < B->GetChildrenSize(); i++)
+        if (A->GetBound().Distance(b_nodes[i]->GetBound()) <= distance)
+            GetNeighbors(A, b_nodes[i], distance, result);
+    } else if (B->IsLeaf()) {
+      a_nodes = A->GetChildrenNodes();
+      for (int i = 0; i < A->GetChildrenSize(); i++)
+        if (a_nodes[i]->GetBound().Distance(B->GetBound()) <= distance)
+            GetNeighbors(a_nodes[i], B, distance, result);
+    } else {
+      a_nodes = A->GetChildrenNodes();
+      b_nodes = B->GetChildrenNodes();
+      a_size = A->GetChildrenSize();
+      b_size = B->GetChildrenSize();
+      bool is_same = A == B;
+      for (int i = 0; i < a_size; i++)
+        for (int j = is_same ? (i) : 0; j < b_size; j++)
+          if (a_nodes[i]->GetBound().Distance(b_nodes[j]->GetBound()) <=
+              distance)
+              GetNeighbors(a_nodes[i], b_nodes[j], distance, result);
+    }
+  }
+}
+
+/**
+ * Neighbour search function
+ * Search pairs (a, b) where a from A and b from B
+ * Absolutelly the same method. The only difference is result type
+ * @tparam T
+ * @param A - node of the tree
+ * @param B - node of the tree
+ * @param distance - finding neighbors within that distance
+ * @param result - container, where we save retrieved data (all neighbor pairs
+ * without its points)
+ */
+template <typename T>
+void spatial_tree_node<T>::GetNeighbors(spatial_tree_node<T> *A,
```

Comment:
Please avoid code duplication:
since the two versions differ only in one line you can use a private internal implementation with an additional parameter. This parameter can be a lambda function which contains the logic for the different result generation
see https://github.com/BioDynaMo/biodynamo/blob/master/demo/cell_growth.cc#L84
let me know if you have questions regarding this